metatensor-torch/include/metatensor/torch/atomistic.hpp

```diff
@@ -0,0 +1,388 @@
+public:
+    /// Create `NeighborsListOptions` with the given `cutoff` and `full_list`.
+    ///
+    /// `requestor` can be used to store information about who requested the
+    /// neighbors list.
+    NeighborsListOptionsHolder(double model_cutoff, bool full_list, std::string requestor = "");
+    ~NeighborsListOptionsHolder() override = default;
+
+    /// Spherical cutoff radius for this neighbors list, in the units of the model
+    double model_cutoff() const {
+        return model_cutoff_;
+    }
+
+    /// Spherical cutoff radius for this neighbors list, in the units of the engine
+    double engine_cutoff() const {
+        return engine_cutoff_;
+    }
+
+    /// Set the conversion factor from the model units to the engine units.
+    ///
+    /// This should be called before `engine_cutoff()`.
+    void set_engine_unit(double conversion) {
+        engine_cutoff_ = model_cutoff_ * conversion;
+    }
+
+    /// Should the list be a full neighbors list (contains both the pair i->j
+    /// and j->i) or a half neighbors list (contains only the pair i->j)
+    bool full_list() const {
+        return full_list_;
+    }
+
+    /// Get the list of strings describing who requested this neighbors list
+    std::vector<std::string> requestors() const {
+        return requestors_;
+    }
+
+    /// Add a new requestor to the list of who requested this neighbors list
+    void add_requestor(std::string requestor);
+
+    /// Implementation of Python's `__repr__`
+    std::string repr() const;
+    /// Implementation of Python's `__str__`
+    std::string str() const;
+
+    /// Serialize a `NeighborsListOptions` to a JSON string.
+    std::string to_json() const;
+    /// Load a serialized `NeighborsListOptions` from a JSON string.
+    static NeighborsListOptions from_json(const std::string& json);
+
+private:
+    // cutoff in the model units
+    double model_cutoff_;
+    // cutoff in the engine units
+    double engine_cutoff_;
+    bool full_list_;
+    std::vector<std::string> requestors_;
+};
+
+/// Check `NeighborsListOptions` for equality. The `requestors` list is ignored
+/// when checking for equality
+inline bool operator==(const NeighborsListOptions& lhs, const NeighborsListOptions& rhs) {
+    return lhs->model_cutoff() == rhs->model_cutoff() && lhs->full_list() == rhs->full_list();
+}
+
+/// Check `NeighborsListOptions` for inequality.
+inline bool operator!=(const NeighborsListOptions& lhs, const NeighborsListOptions& rhs) {
+    return !(lhs == rhs);
+}
+
+
+/// Custom autograd node going (positions, cell) => neighbors list distances
+///
+/// This does not actually computes the neighbors list, but registers a new node
+/// in the overall computational graph, allowing to re-use neighbors list
+/// computed outside of torch. The public interface for this is the
+/// `register_autograd_neighbors()` function.
+class METATENSOR_TORCH_EXPORT NeighborsAutograd: public torch::autograd::Function<NeighborsAutograd> {
```

Comment:
What does this do? Isn't the model supposed to handle this part using torch operations? What use case were you thinking of?

Reply:
>  What use case were you thinking of?

> allowing to re-use neighbors list computed outside of torch.

I.e. use a NL computed in LAMMPS, but pretend that it was computed with pytorch for autograd purposes. This avoid having to re-do the `positions[j] - positions[i] + shift @ cell` calculation, since we already have the corresponding values.

Reply:
Ok, good!